File: internal/files/text_write.go
// text_write.go is the write path for global instruction files (CLAUDE.md,
// RTK.md, rules/, commands/, tools/) that steer every Claude Code session —
// a SECURITY-CRITICAL surface. Every path here funnels through the same
// lexical-clean + confine-PARENT-to-root pipeline as
// internal/maintenance/trash.go's confineParentToRoot: Confine() returns a
// non-existent candidate UNCHANGED with zero containment check (see its own
// doc comment), so guarding the parent — never the not-yet-existing leaf —
// is the only safe way to block a symlinked rules/ escaping root. Do NOT
// weaken this pipeline or bypass it for a new caller.
package files

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"unicode/utf8"

	"claude-devtools/internal/tokenizer"
)

// allowKind discriminates instructionAllowlist entries. allowSegmentGlob
// (e.g. W28's projects/*/memory/**) is a deliberately unbuilt future kind —
// leave the enum extensible, don't add glob matching until it's needed.
type allowKind int

const (
	allowExactFile allowKind = iota
	allowDirPrefix
)

type allowRule struct {
	kind allowKind
	path string
}

// instructionAllowlist is the single, data-driven source of truth for which
// relative paths under the claude root are readable/writable/listable as
// instruction files. Read/write/list all consume it through matchAllowlist
// so the three surfaces can never drift out of sync.
var instructionAllowlist = []allowRule{
	{allowExactFile, "CLAUDE.md"},
	{allowExactFile, "RTK.md"},
	{allowDirPrefix, "rules"},
	{allowDirPrefix, "commands"},
	{allowDirPrefix, "tools"},
}

// matchAllowlist reports whether cleaned (already filepath.Clean'd, relative)
// matches an allowlist entry. Directory-prefix matches are segment-bounded —
// "rules" or "rules/x.md" match, but "rules-evil.md" does NOT — so a
// sibling file crafted to share the "rules" prefix textually is rejected.
func matchAllowlist(cleaned string) bool {
	for _, rule := range instructionAllowlist {
		switch rule.kind {
		case allowExactFile:
			if cleaned == rule.path {
				return true
			}
		case allowDirPrefix:
			if cleaned == rule.path || strings.HasPrefix(cleaned, rule.path+string(filepath.Separator)) {
				return true
			}
		}
	}
	return false
}

// validateRelPath lexically rejects any relPath that isn't already in
// canonical, relative, non-parent-escaping form, THEN checks the allowlist —
// both gates run before any filesystem call.
func validateRelPath(relPath string) (string, error) {
	cleaned := filepath.Clean(relPath)
	if cleaned != relPath || filepath.IsAbs(cleaned) || cleaned == ".." ||
		strings.HasPrefix(cleaned, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("files: invalid instruction file path %q", relPath)
	}
	if !matchAllowlist(cleaned) {
		return "", fmt.Errorf("files: %q is not in the instruction-file allowlist", relPath)
	}
	return cleaned, nil
}

// ResolveInstructionPath validates relPath (lexical + allowlist) and
// resolves it to an absolute path confined within root, following the
// confineParentToRoot pattern: canonicalize root, create relPath's parent
// directory if missing, then EvalSymlinks + Confine the PARENT (never the
// leaf, which may not exist yet) to block a symlinked rules/ pointing
// outside root. Shared by ReadTextFile, MutateTextFile, and
// maintenanceservice's DeleteInstructionFile so all three enforce identical
// safety.
func ResolveInstructionPath(root, relPath string) (string, error) {
	cleaned, err := validateRelPath(relPath)
	if err != nil {
		return "", err
	}

	canonRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return "", fmt.Errorf("files: instruction root %q: %w", root, err)
	}

	abs := filepath.Join(canonRoot, cleaned)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		return "", fmt.Errorf("files: create instruction directory: %w", err)
	}

	parentCanon, err := filepath.EvalSymlinks(filepath.Dir(abs))
	if err != nil {
		return "", fmt.Errorf("files: instruction parent directory: %w", err)
	}
	if _, err := Confine(parentCanon, canonRoot); err != nil {
		return "", err
	}

	return filepath.Join(parentCanon, filepath.Base(cleaned)), nil
}

// instructionWriteMu is the single mutex for the whole instruction-file
// family — one lock, not a per-path sync.Map, since MaintenanceService's own
// s.mu already serializes at the service layer and a per-path map would just
// add complexity with no real concurrency benefit here.
var instructionWriteMu sync.Mutex

// MutateTextFile is the atomic read-transform-write primitive every
// instruction-file mutator (WriteTextFile's blind replace today, a future
// frontmatter-aware patch later) routes through. It locks
// instructionWriteMu, resolves and confines relPath, reads the CURRENT bytes
// fresh (never a caller's stale snapshot — closes the read-fresh-under-lock
// TOCTOU window), runs transform, rejects non-UTF-8 output, backs up the
// previous content to relPath+".bak" (only if it existed), and writes the
// new content — both the .bak and the final content via temp+rename inside
// the symlink-resolved parent, never os.WriteFile+O_TRUNC over a dest that
// could itself be a symlink.
func MutateTextFile(root, relPath string, transform func(current []byte) ([]byte, error)) error {
	instructionWriteMu.Lock()
	defer instructionWriteMu.Unlock()

	dest, err := ResolveInstructionPath(root, relPath)
	if err != nil {
		return err
	}

	current, err := os.ReadFile(dest)
	hadCurrent := err == nil
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("files: read %q: %w", relPath, err)
	}
	if !hadCurrent {
		current = nil
	}

	next, err := transform(current)
	if err != nil {
		return err
	}
	if !utf8.Valid(next) {
		return fmt.Errorf("files: refusing to write non-UTF-8 content to %q", relPath)
	}

	if hadCurrent {
		if err := atomicWriteFile(dest+".bak", current); err != nil {
			return fmt.Errorf("files: write backup for %q: %w", relPath, err)
		}
	}
	if err := atomicWriteFile(dest, next); err != nil {
		return fmt.Errorf("files: write %q: %w", relPath, err)
	}
	return nil
}

// WriteTextFile blindly replaces relPath's whole content — the normal
// editor-save path — through MutateTextFile so it shares the same
// lock/confine/backup/atomic-write pipeline as every other instruction-file
// mutator. No format transformation: content is written byte-for-byte,
// trailing newline and all.
func WriteTextFile(root, relPath string, content []byte) error {
	return MutateTextFile(root, relPath, func(_ []byte) ([]byte, error) {
		if !utf8.Valid(content) {
			return nil, fmt.Errorf("files: refusing to write non-UTF-8 content to %q", relPath)
		}
		return content, nil
	})
}

// ReadTextFile returns relPath's raw bytes, applying the same lexical +
// allowlist + confine-parent-to-root safety as MutateTextFile — a read must
// not escape root either. Non-UTF-8 file content is rejected; the editor is
// UTF-8-text-only.
func ReadTextFile(root, relPath string) ([]byte, error) {
	dest, err := ResolveInstructionPath(root, relPath)
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(dest)
	if err != nil {
		return nil, fmt.Errorf("files: read %q: %w", relPath, err)
	}
	if !utf8.Valid(data) {
		return nil, fmt.Errorf("files: %q is not valid UTF-8", relPath)
	}
	return data, nil
}

// InstructionFile is one entry in ListInstructionFiles' result: an
// allowlisted file's size and approximate context-window cost.
type InstructionFile struct {
	RelPath      string `json:"relPath"`
	Bytes        int    `json:"bytes"`
	ApproxTokens int    `json:"approxTokens"`
}

// ListInstructionFiles walks every instructionAllowlist entry under root and
// reports each matched file's size + token cost (via
// internal/tokenizer.CountTokens, not a bytes/4 estimate) — the
// context-cost meter's data source. Read-only: a missing root or a missing
// individual allowlist directory yields no entries for it, not an error.
func ListInstructionFiles(root string) ([]InstructionFile, error) {
	canonRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return nil, fmt.Errorf("files: instruction root %q: %w", root, err)
	}

	var out []InstructionFile
	for _, rule := range instructionAllowlist {
		switch rule.kind {
		case allowExactFile:
			if entry, ok := statInstructionFile(canonRoot, rule.path); ok {
				out = append(out, entry)
			}
		case allowDirPrefix:
			entries, err := walkInstructionDir(canonRoot, rule.path)
			if err != nil {
				continue // missing/unreadable directory — nothing to list under it yet
			}
			out = append(out, entries...)
		}
	}

	sort.Slice(out, func(i, j int) bool { return out[i].RelPath < out[j].RelPath })
	return out, nil
}

func statInstructionFile(canonRoot, relPath string) (InstructionFile, bool) {
	content, err := os.ReadFile(filepath.Join(canonRoot, relPath))
	if err != nil {
		return InstructionFile{}, false
	}
	return InstructionFile{
		RelPath:      relPath,
		Bytes:        len(content),
		ApproxTokens: tokenizer.CountTokens(string(content)),
	}, true
}

// walkInstructionDir lists every file under canonRoot/dirRelPath, skipping
// subdirectories, the write primitive's own ".bak"/".tmp" byproducts (never
// loaded into a session's context, so counting them would misreport the
// meter), and anything matchAllowlist itself would reject.
func walkInstructionDir(canonRoot, dirRelPath string) ([]InstructionFile, error) {
	dirAbs := filepath.Join(canonRoot, dirRelPath)
	if _, err := os.Stat(dirAbs); err != nil {
		return nil, err
	}

	var out []InstructionFile
	err := filepath.WalkDir(dirAbs, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		if strings.HasSuffix(d.Name(), ".bak") || strings.HasSuffix(d.Name(), ".tmp") {
			return nil
		}
		rel, relErr := filepath.Rel(canonRoot, path)
		if relErr != nil || !matchAllowlist(rel) {
			return nil
		}
		content, readErr := os.ReadFile(path)
		if readErr != nil {
			return nil
		}
		out = append(out, InstructionFile{
			RelPath:      rel,
			Bytes:        len(content),
			ApproxTokens: tokenizer.CountTokens(string(content)),
		})
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

File: src-tauri/src/files/text_write.rs
//! Ports `internal/files/text_write.go` — the write path for global instruction
//! files (CLAUDE.md, RTK.md, rules/, commands/, tools/) that steer every Claude
//! Code session — a SECURITY-CRITICAL surface. Every path funnels through the
//! same lexical-clean + allowlist + confine-PARENT-to-root pipeline:
//! `pathutil::confine` returns a non-existent candidate UNCHANGED with zero
//! containment check, so guarding the PARENT — never the not-yet-existing leaf —
//! is the only safe way to block a symlinked `rules/` escaping root.
//!
//! `clean` and `atomic_write_file` are reused from `agents_write.rs` (Go shares
//! `filepath.Clean` / `atomicWriteFile` package-wide).

use std::fs;
use std::path::Path;
use std::sync::{LazyLock, Mutex};

use serde::{Deserialize, Serialize};

use crate::files::agents_write::{atomic_write_file, clean};
use crate::files::fsutil;
use crate::files::pathutil::confine;
use crate::tokenizer::count_tokens;

/// Discriminates `INSTRUCTION_ALLOWLIST` entries. A future segment-glob kind
/// (e.g. `projects/*/memory/**`) is deliberately unbuilt — the enum stays
/// extensible, glob matching isn't added until needed. Mirrors `allowKind`.
enum AllowKind {
    ExactFile,
    DirPrefix,
}

struct AllowRule {
    kind: AllowKind,
    path: &'static str,
}

/// The single, data-driven source of truth for which relative paths under the
/// claude root are readable/writable/listable. Read/write/list all consume it
/// through `match_allowlist` so the three surfaces can never drift.
static INSTRUCTION_ALLOWLIST: &[AllowRule] = &[
    AllowRule {
        kind: AllowKind::ExactFile,
        path: "CLAUDE.md",
    },
    AllowRule {
        kind: AllowKind::ExactFile,
        path: "RTK.md",
    },
    AllowRule {
        kind: AllowKind::DirPrefix,
        path: "rules",
    },
    AllowRule {
        kind: AllowKind::DirPrefix,
        path: "commands",
    },
    AllowRule {
        kind: AllowKind::DirPrefix,
        path: "tools",
    },
];

/// The single mutex for the whole instruction-file family — one lock, not a
/// per-path map — since the service layer already serializes. Mirrors
/// `instructionWriteMu`.
static INSTRUCTION_WRITE_MU: LazyLock<Mutex<()>> = LazyLock::new(|| Mutex::new(()));

/// Reports whether `cleaned` (already `clean`'d, relative) matches an allowlist
/// entry. Directory-prefix matches are segment-bounded — `rules`/`rules/x.md`
/// match, but `rules-evil.md` does NOT. Mirrors `matchAllowlist`.
fn match_allowlist(cleaned: &str) -> bool {
    for rule in INSTRUCTION_ALLOWLIST {
        match rule.kind {
            AllowKind::ExactFile => {
                if cleaned == rule.path {
                    return true;
                }
            }
            AllowKind::DirPrefix => {
                if cleaned == rule.path || cleaned.starts_with(&format!("{}/", rule.path)) {
                    return true;
                }
            }
        }
    }
    false
}

/// Lexically rejects any `rel_path` not already in canonical, relative,
/// non-parent-escaping form, THEN checks the allowlist — both gates before any
/// filesystem call. Mirrors `validateRelPath`; error sentinels byte-identical.
fn validate_rel_path(rel_path: &str) -> Result<String, String> {
    let cleaned = clean(rel_path);
    if cleaned != rel_path
        || Path::new(&cleaned).is_absolute()
        || cleaned == ".."
        || cleaned.starts_with("../")
    {
        return Err(format!("files: invalid instruction file path {rel_path:?}"));
    }
    if !match_allowlist(&cleaned) {
        return Err(format!(
            "files: {rel_path:?} is not in the instruction-file allowlist"
        ));
    }
    Ok(cleaned)
}

/// Validates `rel_path` (lexical + allowlist) and resolves it to an absolute path
/// confined within `root`: canonicalize root, create the leaf's parent if
/// missing, then canonicalize + confine the PARENT (never the leaf). Mirrors
/// `ResolveInstructionPath`.
pub fn resolve_instruction_path(root: &str, rel_path: &str) -> Result<String, String> {
    let cleaned = validate_rel_path(rel_path)?;

    let canon_root =
        fs::canonicalize(root).map_err(|e| format!("files: instruction root {root:?}: {e}"))?;

    let abs = canon_root.join(&cleaned);
    let parent = abs.parent().unwrap_or(canon_root.as_path());
    fs::create_dir_all(parent).map_err(|e| format!("files: create instruction directory: {e}"))?;

    let parent_canon =
        fs::canonicalize(parent).map_err(|e| format!("files: instruction parent directory: {e}"))?;
    confine(
        &parent_canon.to_string_lossy(),
        &canon_root.to_string_lossy(),
    )?;

    let base = Path::new(&cleaned).file_name().unwrap_or_default();
    Ok(parent_canon.join(base).to_string_lossy().into_owned())
}

/// The atomic read-transform-write primitive every instruction-file mutator
/// routes through: lock, resolve+confine, read the CURRENT bytes fresh (closes
/// the TOCTOU window), run `transform`, reject non-UTF-8 output, back up the
/// previous content to `rel_path.bak` (only if it existed), write the new
/// content — both `.bak` and final via temp+rename. Mirrors `MutateTextFile`.
pub fn mutate_text_file<F>(root: &str, rel_path: &str, transform: F) -> Result<(), String>
where
    F: FnOnce(&[u8]) -> Result<Vec<u8>, String>,
{
    let _guard = fsutil::lock(&INSTRUCTION_WRITE_MU);

    let dest = resolve_instruction_path(root, rel_path)?;

    let (current, had_current): (Vec<u8>, bool) = match fs::read(&dest) {
        Ok(bytes) => (bytes, true),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => (Vec::new(), false),
        Err(e) => return Err(format!("files: read {rel_path:?}: {e}")),
    };

    let next = transform(&current)?;
    if std::str::from_utf8(&next).is_err() {
        return Err(format!(
            "files: refusing to write non-UTF-8 content to {rel_path:?}"
        ));
    }

    if had_current {
        atomic_write_file(&format!("{dest}.bak"), &current)
            .map_err(|e| format!("files: write backup for {rel_path:?}: {e}"))?;
    }
    atomic_write_file(&dest, &next).map_err(|e| format!("files: write {rel_path:?}: {e}"))?;
    Ok(())
}

/// Blindly replaces `rel_path`'s whole content — the normal editor-save path —
/// through `mutate_text_file`. No format transformation: content is written
/// byte-for-byte. Mirrors `WriteTextFile`.
pub fn write_text_file(root: &str, rel_path: &str, content: &[u8]) -> Result<(), String> {
    mutate_text_file(root, rel_path, |_current| {
        if std::str::from_utf8(content).is_err() {
            return Err(format!(
                "files: refusing to write non-UTF-8 content to {rel_path:?}"
            ));
        }
        Ok(content.to_vec())
    })
}

/// Returns `rel_path`'s raw bytes, applying the same lexical + allowlist +
/// confine-parent safety as `mutate_text_file`. Non-UTF-8 content is rejected.
/// Mirrors `ReadTextFile`.
pub fn read_text_file(root: &str, rel_path: &str) -> Result<Vec<u8>, String> {
    let dest = resolve_instruction_path(root, rel_path)?;
    let data = fs::read(&dest).map_err(|e| format!("files: read {rel_path:?}: {e}"))?;
    if std::str::from_utf8(&data).is_err() {
        return Err(format!("files: {rel_path:?} is not valid UTF-8"));
    }
    Ok(data)
}

/// One entry in `list_instruction_files`' result: an allowlisted file's size and
/// approximate context-window cost. Mirrors `InstructionFile`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InstructionFile {
    pub rel_path: String,
    pub bytes: usize,
    pub approx_tokens: usize,
}

/// Walks every `INSTRUCTION_ALLOWLIST` entry under `root` and reports each
/// matched file's size + token cost (via `tokenizer::count_tokens`, not a
/// bytes/4 estimate). Read-only: a missing root or individual dir yields no
/// entries for it, not an error. Mirrors `ListInstructionFiles`.
pub fn list_instruction_files(root: &str) -> Result<Vec<InstructionFile>, String> {
    let canon_root =
        fs::canonicalize(root).map_err(|e| format!("files: instruction root {root:?}: {e}"))?;

    let mut out: Vec<InstructionFile> = Vec::new();
    for rule in INSTRUCTION_ALLOWLIST {
        match rule.kind {
            AllowKind::ExactFile => {
                if let Some(entry) = stat_instruction_file(&canon_root, rule.path) {
                    out.push(entry);
                }
            }
            AllowKind::DirPrefix => match walk_instruction_dir(&canon_root, rule.path) {
                Ok(entries) => out.extend(entries),
                Err(_) => continue,
            },
        }
    }

    out.sort_by(|a, b| a.rel_path.cmp(&b.rel_path));
    Ok(out)
}

/// Mirrors `statInstructionFile`.
fn stat_instruction_file(canon_root: &Path, rel_path: &str) -> Option<InstructionFile> {
    let content = fs::read(canon_root.join(rel_path)).ok()?;
    Some(InstructionFile {
        rel_path: rel_path.to_string(),
        bytes: content.len(),
        approx_tokens: count_tokens(&String::from_utf8_lossy(&content)),
    })
}

/// Lists every file under `canon_root/dir_rel_path`, skipping subdirectories, the
/// write primitive's own `.bak`/`.tmp` byproducts, and anything
/// `match_allowlist` would reject. Mirrors `walkInstructionDir`.
fn walk_instruction_dir(canon_root: &Path, dir_rel_path: &str) -> Result<Vec<InstructionFile>, String> {
    let dir_abs = canon_root.join(dir_rel_path);
    fs::metadata(&dir_abs).map_err(|e| e.to_string())?;

    let mut out: Vec<InstructionFile> = Vec::new();
    walk_dir_recursive(canon_root, &dir_abs, &mut out)?;
    Ok(out)
}

fn walk_dir_recursive(
    canon_root: &Path,
    dir: &Path,
    out: &mut Vec<InstructionFile>,
) -> Result<(), String> {
    let entries = fs::read_dir(dir).map_err(|e| e.to_string())?;
    for entry in entries {
        let entry = entry.map_err(|e| e.to_string())?;
        let file_type = entry.file_type().map_err(|e| e.to_string())?;
        let path = entry.path();
        if file_type.is_dir() {
            walk_dir_recursive(canon_root, &path, out)?;
            continue;
        }
        let name = entry.file_name().to_string_lossy().into_owned();
        if name.ends_with(".bak") || name.ends_with(".tmp") {
            continue;
        }
        let rel = match path.strip_prefix(canon_root) {
            Ok(r) => r.to_string_lossy().into_owned(),
            Err(_) => continue,
        };
        if !match_allowlist(&rel) {
            continue;
        }
        let content = match fs::read(&path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        out.push(InstructionFile {
            rel_path: rel,
            bytes: content.len(),
            approx_tokens: count_tokens(&String::from_utf8_lossy(&content)),
        });
    }
    Ok(())
}

#[cfg(test)]
#[path = "text_write_tests.rs"]
mod text_write_tests;
